libsql-server/src/connection/mod.rs

```diff
@@ -189,6 +197,89 @@ where
     }
 }
 
+pub struct MakeSharedSchemaConnection<F> {
+    inner: F,
+    is_shared_schema: bool,
+}
+
+impl<F> MakeSharedSchemaConnection<F> {
+    pub fn new(inner: F, is_shared_schema: bool) -> Self {
+        MakeSharedSchemaConnection {
+            inner,
+            is_shared_schema,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<F: MakeConnection> MakeConnection for MakeSharedSchemaConnection<F> {
+    type Connection = PrimaryConnection<F::Connection>;
+
+    async fn create(&self) -> Result<Self::Connection, Error> {
+        let conn = self.inner.create().await?;
+        Ok(if self.is_shared_schema {
+            PrimaryConnection::SharedSchema(SharedSchemaConnection::new(conn))
+        } else {
+            PrimaryConnection::Tracked(conn)
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct SharedSchemaConnection<DB> {
+    inner: DB,
+}
+
+impl<DB> SharedSchemaConnection<DB> {
+    fn new(inner: DB) -> Self {
+        SharedSchemaConnection { inner }
+    }
+}
+
+#[async_trait::async_trait]
+impl<DB: Connection> Connection for SharedSchemaConnection<DB> {
+    async fn execute_program<B: QueryResultBuilder>(
+        &self,
+        pgm: Program,
+        auth: Authenticated,
+        response_builder: B,
+        replication_index: Option<FrameNo>,
+    ) -> Result<B> {
+        //TODO: execute multi-database execution procedure
+        todo!()
```

Comment:
This part will be responsible for executing multi-database execution. We'll need to modify current connection state with data necessary for driving the execution process, dry_runs etc.